pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1393,6 +1452,415 @@ public PartitionAssignment getStreamPartitionAssignment(String realtimeTableName
+        if (prevMetadata != null) {
+          prevSegmentId = prevMetadata.getSegmentName();
+        }
+        if (skipNewPartitions && prevMetadata == null) {
+          continue;
+        }
+
+        segmentMapper.addMapping(prevSegmentId, segmentId);
+      }
+    }
+
+    if (!skipNewPartitions) {
+      for (int partition : newPartitions) {
+        int nextSeqNum = STARTING_SEQUENCE_NUMBER;
+        LOGGER.info("Creating CONSUMING segment for {} partition {} with seq {}", tableNameWithType, partition,
+            nextSeqNum);
+        String consumerStartOffsetSpec = streamMetadata.getKafkaConsumerProperties().get(CommonConstants.Helix.DataSource.Realtime.Kafka.AUTO_OFFSET_RESET);
+        long startOffset = getKafkaPartitionOffset(streamMetadata, consumerStartOffsetSpec, partition);
+        LOGGER.info("Found kafka offset {} for table {} for partition {}", startOffset, tableNameWithType, partition);
+        String rawTableName = TableNameBuilder.extractRawTableName(tableNameWithType);
+
+        LLCSegmentName newLLCSegmentName = new LLCSegmentName(rawTableName, partition, nextSeqNum, now);
+
+        createNewSegmentMetadataZNRecord(tableNameWithType, newLLCSegmentName, startOffset, partitionAssignment);
+        segmentMapper.addMapping(segmentMapper.generateNewSegmentKey(partition), newLLCSegmentName.getSegmentName());
+      }
+    }
+
+    RealtimeSegmentAssignmentStrategy segmentAssignmentStrategy = new ConsumingSegmentAssignmentStrategy();
+    Map<String, List<String>> assignments = null;
+    try {
+      assignments =
+          segmentAssignmentStrategy.assign(Lists.newArrayList(segmentMapper.getAllNewSegmentNames()), partitionAssignment);
+    } catch (InvalidConfigException e) {
+      throw new IllegalStateException(
+          "Caught exception when assigning segments using partition assignment for table " + tableNameWithType);
+    }
+
+    for (Map.Entry<String, String> entry : segmentMapper.getAll().entrySet()) {
+      String oldSeg = entry.getKey();
+      String newSeg = entry.getValue();
+      List<String> newInstances = assignments.get(newSeg);
+      Map<String, String> instanceStateMap = idealState.getInstanceStateMap(newSeg);
+      if (instanceStateMap != null) {
+        instanceStateMap.clear();
+      }
+      for (String instance : newInstances) {
+        idealState.setPartitionState(newSeg, instance,
+            PinotHelixSegmentOnlineOfflineStateModelGenerator.CONSUMING_STATE);
+      }
+      if (!segmentMapper.isNewSegmentKey(oldSeg)) {
+        Set<String> oldInstances = idealState.getInstanceSet(oldSeg);
+        Preconditions.checkArgument(CollectionUtils.isNotEmpty(oldInstances));
+        for (String instance : oldInstances) {
+          idealState.setPartitionState(oldSeg, instance,
+              PinotHelixSegmentOnlineOfflineStateModelGenerator.ONLINE_STATE);
+        }
+      }
+    }
+    return idealState;
+  }
```

Comment:
we are missing the logic to update the nonconsumingPartitionCouintMetric (in validationmanager), but i think that metric was not updated correctly anyway. We need to create a new controller guage that reflects this. We can do that independently.

Reply:
Added a TODO for this